src/dune_rules/coq_lib.ml

```diff
@@ -38,128 +67,333 @@ let libraries l = l.libraries
+      (entries : (Coq_stanza.Theory.t * entry) list) =
+    let t = Fdecl.create Dyn.opaque in
     let boot =
-      match List.find_all ~f:(fun (_, s) -> s.Coq_stanza.Theory.boot) sl with
-      | [] -> None
-      | [ l ] -> Some ((snd l).buildable.loc, snd (create_from_stanza l))
-      | _ :: (_, s2) :: _ ->
-        Result.ok_exn (Error.duplicate_boot_lib ~loc:s2.buildable.loc s2)
+      let boot =
+        match
+          List.find_all entries
+            ~f:(fun ((theory : Coq_stanza.Theory.t), _entry) -> theory.boot)
+        with
+        | [] -> None
+        | [ ((theory : Coq_stanza.Theory.t), _entry) ] ->
+          let loc, name = theory.name in
+          Some (loc, name, theory.buildable.coq_lang_version)
+        | boots ->
+          let stanzas = List.map boots ~f:fst in
+          Error.duplicate_boot_lib stanzas
+      in
+      match boot with
+      | None -> None
+      | Some (loc, name, coq_lang_version) ->
+        let lib =
+          Memo.lazy_ (fun () ->
+              let t = Fdecl.get t in
+              resolve t ~coq_lang_version (loc, name))
+        in
+        Some (loc, lib)
     in
-    { boot; libs }
-
-  let resolve ?(allow_private_deps = true) db (loc, name) =
-    match Coq_lib_name.Map.find db.libs name with
-    | Some s ->
-      if (not allow_private_deps) && Option.is_none s.package then
-        Error.private_deps_not_allowed ~loc s
-      else Ok s
-    | None -> Error.theory_not_found ~loc name
-
-  let find_many t ~loc = Result.List.map ~f:(fun name -> resolve t (loc, name))
-
-  module Coq_lib_closure = Top_closure.Make (String.Set) (Or_exn)
-
-  let add_boot db theories =
-    match db.boot with
-    | None -> theories
-    (* XXX: Note that this means that we will prefix Coq with -Q, not sure we
-       want to do that (yet), but seems like good practice. *)
-    | Some (loc, stdlib) -> (loc, snd stdlib.name) :: theories
-
-  let top_closure db theories ~allow_private_deps ~loc =
-    let open Result.O in
-    let* theories =
-      add_boot db theories
-      |> Result.List.map ~f:(resolve ~allow_private_deps db)
+    let resolve =
+      let map =
+        match
+          Coq_lib_name.Map.of_list_map entries
+            ~f:(fun ((theory : Coq_stanza.Theory.t), entry) ->
+              (snd theory.name, (theory, entry)))
+        with
+        | Ok m -> m
+        | Error (_name, _, (theory, _entry)) ->
+          Error.duplicate_theory_name theory.name
+      in
+      fun name ->
+        match Coq_lib_name.Map.find map name with
+        | None -> `Not_found
+        | Some (theory, entry) -> (
+          match entry with
+          | Theory dir -> `Theory (find_db dir, dir, theory)
+          | Redirect db -> `Redirect db)
     in
-    let key (t : lib) = Coq_lib_name.to_string (snd t.name) in
-    let deps (t : lib) =
-      Result.List.map ~f:(resolve ~allow_private_deps db) t.theories
+    Fdecl.set t { boot; resolve; parent };
+    Fdecl.get t
```

Comment:
This game with `Fdecl` always bothered me a bit; I guess that's Ok but I'd be nice if we could write code here in direct-style. At some point reasoning about what is going on becomes tricky.

Reply:
Unfortunately I don't see how to get rid of it. To resolve the boot library we need the db, and to resolve all the other libs in the db, we need the boot library. When stdlib is a normal lib, we can drop this junk.

Reply:
Thanks, let me think about it for a sec.